### DISCUSSION TITLE
Bounce Rate and Visit Duration

### DISCUSSION THREAD

**Discussion by harm0nic** [plausible/analytics#863] [Anything Else]
As per the metrics definitions:

**Bounce Rate**
The percentage of visitors with a single page view. A visitor "bounces" away and leaves your site after only viewing a single page.

**Visit Duration**
How long time visitors spend on your site. It only shows people who visit more than one page. For those who visit one page only we default to 0 seconds.

My concern is that several of my sites are essentially 1-page sites, so using plausible, I'd have a bounce rate of 100% and a visit duration one 0.

From looking around here, I gather that the reasoning that plausible works like this is because they are adhering strickly to the definition of what a bounce is. The problem is that this strict definition actively hurts the data for use cases like single-page sites. I'm not too concerned about the Bounce Rate data for a single-page site (since in that case, 100% and 0% are equally useless), but the inability to capture visit duration is a considerable issue here.

Is there any intention to perhaps have a setting where we can soften the rule on when to capture duration data? If not, what is the reasoning behind this decision? Is there some technical limitation where we can only capture duration data once an internal link has been clicked, but "exit" pages cannot be captured?

**Comment by metmarkosaric** [comment]
thanks @harm0nic! this is the typical definition of these metrics even on Google Analytics. but Plausible supports single page applications out of the box so both bounce rate and visit duration will be calculated correctly. see https://plausible.io/docs/spa-support. and if your single page application uses hash based routing, we have a special script you should use to do the same: https://plausible.io/docs/hash-based-routing

**Reply by harm0nic** [reply]
First, thanks for the response Marko!

Second, for my example, I'm not talking about an SPA with routing or anything like that. Like, just a single regular HTML page. No routing, no react or anything. Just a plain old boring webpage. It is my understanding that Plausible will always return the visit duration of a site like this as zero.

**Reply by metmarkosaric** [reply]
I see. Did you try to setup any custom events on the page? Triggering a custom event after a pageview will mark the session as a non-bounce

**Comment by ukutaht** [comment]
Yes this is a limitation for your use-case. You can try `setTimeout(sendCustomEvent, 15 * 1000)` or something like that to mark a 15-second session as not bounced. However, these custom events will count towards your billable pageviews.

> Is there any intention to perhaps have a setting where we can soften the rule on when to capture duration data?

Maybe in the future we can do something with `sendBeacon` to get end-of-pageview timing. This would give more accurate time on page, visit duration and would allow a more flexible definition of engaged (not bounced) visitors.

> If not, what is the reasoning behind this decision?

Lack of bandwidth. I am one person doing pretty much all development (with help from great contributors like @Vigasaurus ). We have feature requests with many more people waiting. You're free to contribute this feature but you'd need quite a lot of knowledge about the system to do it.

**Reply by harm0nic** [reply]
Just looked up `sendBeacon` it looks like implementation of it would be amazing! Consider this a vote in favour :+1: 

Other than that, sending a custom event seems like a good option - for now at least. Using 15s as an example, if a user is on the page for longer than that, then it's a good indicator that when they leave it is not a "real" bounce. Thanks for the idea.

**Reply by ukutaht** [reply]
Cheers, thanks for understanding 🙏

**Reply by fabiocaccamo** [reply]
I agree with the timeout idea, perhaps it would be nice to have the possibility to configure it in the website configuration options (or as data attribute) without having to write any js for it.

**Comment by changethe** [comment]
**Regarding the bouncerate:**
What constitutes a bounce for me differs from the standard definition of “only visited one page”. I define a visit that lasted at least 20 seconds (even if it was just only a single page visit) as not bounced. I can easily fix this by setting a ping via custom event that triggers after a 20sec timeout. As @harm0nic  mentioned, it might be handy to have this as a setting in the site settings (something along the lines of “bounce rate timer”), but i guess overall it’s not a big issue and is easily solvable as described above.

**Regarding visit duration and time on page:**
These metrics are fundamentally flawed and pretty much useless currently in my opinion. 
The way they are currently measured means that:
1. times are only tracked between events, which is very inaccurate
2. times are polluted by inactive time

Simpleanalytics has a good article on this: [https://docs.simpleanalytics.com/explained/time-on-page](https://docs.simpleanalytics.com/explained/time-on-page). And while I don’t fully agree with using median instead of mean for the totals, the other things are very important points.


**My suggestion:**
Remove the “visit duration” and “time on page” metrics and replace them with an “engaged time” metric, much along the points made in https://github.com/plausible/analytics/discussions/591

To gather this data properly, the collection method probably will have to be switched to using the beacon api, as stated in https://github.com/plausible/analytics/discussions/190

Some inspiration regarding this:

- https://github.com/jasonzissman/TimeMe.js
This library is very bloated, especially since I feel like the script should stay as slim as possible, but I think there might be some good ideas to be gained from how they calculate the engagement time.
And if it ends up needing too much code, it could be offered only as an optional add-on to the script, much like the current “track outbound links” option.

- https://github.com/umami-software/umami/issues/1470 and https://github.com/umami-software/umami/issues/1160
Interesting discussion around the caveats regarding the beacon api.
